### PR TITLE
Add support for subtests

### LINKED ISSUE #71: Sub tests are not included
If the unittests include [subtests](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests) the results of these subtests are not picked up.

Also, if the test class contains only failing subtests, the tap file may not be generated at all.

### PR DESCRIPTION
**Pull request by ecederstrand**
Adds support for subtests.

Fixes #71

To accept your contribution, please ensure that the checklist below is complete.

* [x] Is your name/identity in the AUTHORS file?
* [x] Does the code change (if the PR contains code) have 100% test coverage?
* [x] Is CI passing all quality and testing checks?

### CODE REVIEW

**Comment by ecederstrand** [comment]
I think coverage may be run on Python 2 where subtests aren't supported. I'm definitely hitting the new `addSubTest()` method in my local tests. Should I make the exposure of that method conditional? E.g.:

```python
class TAPTestResult(TextTestResult):
    if sys.version_info[0] >= 3:
        def addSubTest(self, test, subtest, err):
            ...
```

**Comment by ecederstrand** [comment]
Aslo, I believe I'm innocent of this test failure: https://travis-ci.org/python-tap/tappy/jobs/622716225?utm_medium=notification&utm_source=github_status

**Comment by mblayman** [comment]
Hi, @ecederstrand, thanks for the contribution! tappy is generally pretty stable so it's been a while since I kicked CI's tires.

Since Python 2 is EOL at the end of this month, this PR prompted me to start up another PR that will drop Python 2 support before the next release. I should get that merged tonight and get the macOS build working so you can merge master back into your PR and get a green build.

I'll give the code a review after Travis CI is happy. Thanks!

File: tap/runner.py [hunk 1]
```diff
@@ -17,6 +17,18 @@ class TAPTestResult(TextTestResult):
     def __init__(self, stream, descriptions, verbosity):
         super(TAPTestResult, self).__init__(stream, descriptions, verbosity)
 
+    def addSubTest(self, test, subtest, err):  # pragma: no cover
```
Comment (mblayman) [inline tap/runner.py]: Let's remove the `pragma: no cover` since coverage is now testing Python 3.

File: tap/tests/test_result.py [hunk 2]
```diff
@@ -70,3 +81,28 @@ def test_adds_unexpected_success(self):
         self.assertEqual(
             line.directive.text, "TODO {}".format(_("(unexpected success)"))
         )
+
+    def test_adds_subtest_success(self):
+        """Test that the runner handles subtest success results."""
+        if sys.version_info[0] < 3:
```
Comment (mblayman) [inline tap/tests/test_result.py]: The version checks in these two tests can be removed now that Python2 is out.

**Comment by mblayman** [comment]
Oh! I forgot to add that I pushed a couple of commits to your branch to save you the trouble of merging master. You'll need to pull locally (my apologies if you're a Git pro and know all this already. It's hard with first time contributor PRs to know ability level).

**Comment by ecederstrand** [comment]
No problem, and thanks for the help with the Python 2 cleanup :-)

Subtests are a way to specify that parts of a test case should be a test case in its own. It's usually done for a single assert in a larger test case:

```python
# test.py
import unittest

from tap import TAPTestRunner


class TestCase(unittest.TestCase):
    def test_subs(self):
        for i in range(4):
            with self.subTest(i=i):
                self.assertEqual(i % 2, 0)


if __name__ == '__main__':
    runner = TAPTestRunner()
    runner.set_stream(True)
    unittest.main(testRunner=runner)
```

And here's an example of the output with this PR:

```bash
$ python test.py 
# TAP results for TestCase
ok 1 test_subs (__main__.TestCase) (i=0)
not ok 2 test_subs (__main__.TestCase) (i=1)
# Traceback (most recent call last):
#   File "/opt/python35/lib/python3.5/unittest/case.py", line 59, in testPartExecutor
#     yield
#   File "/opt/python35/lib/python3.5/unittest/case.py", line 523, in subTest
#     yield
#   File "test.py", line 10, in test_subs
#     self.assertEqual(i % 2, 0)
#   File "/opt/python35/lib/python3.5/unittest/case.py", line 829, in assertEqual
#     assertion_func(first, second, msg=msg)
#   File "/opt/python35/lib/python3.5/unittest/case.py", line 822, in _baseAssertEqual
#     raise self.failureException(msg)
# AssertionError: 1 != 0
ok 3 test_subs (__main__.TestCase) (i=2)
not ok 4 test_subs (__main__.TestCase) (i=3)
# Traceback (most recent call last):
#   File "/opt/python35/lib/python3.5/unittest/case.py", line 59, in testPartExecutor
#     yield
#   File "/opt/python35/lib/python3.5/unittest/case.py", line 523, in subTest
#     yield
#   File "test.py", line 10, in test_subs
#     self.assertEqual(i % 2, 0)
#   File "/opt/python35/lib/python3.5/unittest/case.py", line 829, in assertEqual
#     assertion_func(first, second, msg=msg)
#   File "/opt/python35/lib/python3.5/unittest/case.py", line 822, in _baseAssertEqual
#     raise self.failureException(msg)
# AssertionError: 1 != 0
1..4
```

**Comment by ecederstrand** [comment]
I believe all requested changes are done now.

I noticed one more usage of `sys.version_info` in tap/tests/test_parser.py but left it untouched to not pollute this PR.

**Comment by ecederstrand** [comment]
Oh, and Python 3.4 is also EOL: https://devguide.python.org/devcycle/#end-of-life-branches

You may want to remove it from travis.yml

File: tap/tests/test_result.py [hunk 3]
```diff
@@ -70,3 +80,24 @@ def test_adds_unexpected_success(self):
...
+        """Test that the runner handles subtest failure results."""
+        result = self._make_one()
+        # Python 3 does some extra testing in unittest on exceptions so fake
+        # the cause as if it were raised.
+        ex = Exception()
+        ex.__cause__ = None
+        test = FakeTestCase()
+        with test.subTest():
+            result.addSubTest(test, test._subtest, (ex.__class__, ex, None))
+        self.assertEqual(len(result.tracker._test_cases["FakeTestCase"]), 1)
```
Comment (mblayman) [inline tap/tests/test_result.py]: Do you mind switching this assert to check that the test line is a "not ok" line? I think that would be a more accurate assert for the spirit of this test.
```suggestion
        line = result.tracker._test_cases["FakeTestCase"][0]
        self.assertFalse(line.ok)
```

**Pull request merged**